File: problems/w4/shorter-than-x.js
As we saw yesterday, JavaScript does not have list comprehensions like Python, but it does have an Array method called filter, which works like the if in the Python list comprehension. Here we'll duplicate the filter examples from the previous question, but with the JavaScript filter method.

// Here's an example that selects all of the numbers from a list that are greater than 4.

// numbers = [1,2,3,4,5,6,7,8]

// greaterThan4 = numbers.filter(function(number) { return number > 4; });

// console.log(greaterThan4)  # --> [5,6,7,8]
// This example will select all of the strings that represent a number:

// strings = ["cat", "1", "a", "5.3"]

// numberStrings = strings.filter(function(string) {
//     return ! isNaN(parseFloat(string));
// });

// console.log(numberStrings)  # --> ["1", "5.3"]
// Implement the function shorterThanX(strings, maxLength) in JavaScript using the filter function.

const shorterThanX = function(strings, maxLength) {
    // your code here
    return strings.filter(function(string) {
        return string.length < maxLength;
    })
};

File: problems/w4/shorter-than-x.py
# This pattern of selecting some of the items from an input list is so common that the ability to do so is built into Python's list comprehensions.

# Here's an example that selects all of the numbers from a list that are greater than 4.

# numbers = [1,2,3,4,5,6,7,8]

# greater_than_4 = [number for number in numbers if number > 4]

# print(greater_than_4)  # --> [5,6,7,8]
# The if number > 4 at the end is what does the filtering.

# This example will select all of the strings that represent a number:

# strings = ["cat", "1", "a", "5.3"]

# number_strings = [string for string in strings if string.isnumeric()]

# print(number_strings)  # --> ["1", "5.3"]
# Here if string.isnumeric() is the part that's specifying which elements should be included.

# Implement the function shorter_than(strings, max_length) in Python using a list comprehension.

def shorter_than_X(strings, max_length):
    # your list comprehension here
    return [i for i in strings if len(i) < max_length]
